packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+                assert self.in_queue is not None, "Input queue not initialized."
+                assert self.loop is not None, "Loop not initialized."
+                # Ok to put these Envelopes in_queue during HTTPChannel initialization?
+                asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
+            except ValueError:
+                logger.error(f"Bad formatted path: {path_name}")
+            except Exception as e:
+                logger.error(f"Error when processing a path. Message: {str(e)}")
+
+    def _decode_path(self, path_name: str, path_value: dict, separator: str = "/"):
+        """Path --> Envelope
+
+        Convert specified API path name and value into in_queue'd Envelope.
+        Envelope's msg consists of the response options for each REST req in path.
+        """
+        path_split = path_name.split(separator)
+
+        to = path_split[1].strip()  # to: name of Skill (petstoresim)?
+        sender = "TBD"  # hardcoded for now
+        protocol_id = PublicId.from_str("fetchai/http:0.1.0")  # hardcoded for now
+        uri = URI(f"http://{self.host}:{self.port}{path_name}")
+        context = EnvelopeContext(connection_id=self.connection_id, uri=uri)
+        # Msg consists of all of the responses for each REST req in JSON format.
+        responses = {}
+        for req_type, req_value in path_value.items():
+            responses[req_type] = req_value["responses"]
+        msg_bytes = json.dumps(responses).encode()
+
+        return Envelope(
+            to=to,
+            sender=sender,
+            protocol_id=protocol_id,
+            message=msg_bytes,
+            context=context,
+        )
+
+    def _process_request(
+        self, http_method: str, url: str, param: str, body: str = "{}"
+    ):
+        """Process incoming API request by packaging into Envelope and sending it in-queue.
+
+        """
+        self._client_id = uuid4().hex
+        protocol_id = PublicId.from_str("fetchai/http:0.1.0")
+        uri = URI(f"http://{self.host}:{self.port}{url}")
+        context = EnvelopeContext(connection_id=self.connection_id, uri=uri)
+        # Prepare the Envelope's message body and encode it into bytes.
+        msg = {
+            "performative": http_method,
+            "path": url,
+            "params": param,
+            "payload": body,
+        }
+        msg_bytes = json.dumps(msg).encode()
+        # Prepare the Envelope itself using the provided variables.
+        envelope = Envelope(
+            to=self.address,
+            sender=self._client_id,
+            protocol_id=protocol_id,
+            context=context,
+            message=msg_bytes,
+        )
+        # Send the Envelope to the Agent's InBox.
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
```

Comment:
You need to get the correct response envelope here as well, unpack it and return its payload etc.

Reply:
Will try that, I'll probably have questions depending on the implementation structure.

Reply:
Update, I currently have the **APIVerificationHelper** (see below) that directly returns the response as it is. Maybe next iteration it will return the response in an Envelope format.

Reply:
As discussed offline, the APIVerificationHelper does not deal with the envelopes.